sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala

```diff
@@ -1705,6 +1705,91 @@ class AdaptiveQueryExecSuite
     }
   }
 
+  test("SPARK-34980: Support coalesce partition through union") {
+    def checkResultPartition(
+        df: Dataset[Row],
+        unionNumber: Int,
+        shuffleReaderNumber: Int,
+        partitionNumber: Int): Unit = {
+      df.collect()
+      assert(collect(df.queryExecution.executedPlan) {
+        case u: UnionExec => u
+      }.size == unionNumber)
+      assert(collect(df.queryExecution.executedPlan) {
+        case r: AQEShuffleReadExec => r
+      }.size === shuffleReaderNumber)
+      assert(df.rdd.partitions.length === partitionNumber)
+    }
+
+    Seq(true, false).foreach { combineUnionEnabled =>
+      val combineUnionConfig = if (combineUnionEnabled) {
+        "" -> ""
+      } else {
+        SQLConf.OPTIMIZER_EXCLUDED_RULES.key ->
+          "org.apache.spark.sql.catalyst.optimizer.CombineUnions"
```

Comment:
does this really matter for the "coalesce through union" feature? I think we can just test the default case, which means this rule is enabled.